src/Lecture3.hs

```diff
@@ -102,86 +101,104 @@ have a lawful 'Monoid' instance.
 
 -}
 
+-- NOTE (Student) It is easier to derive num as Gold behaves like a
+-- simple Int anyway
 newtype Gold = Gold
-    { unGold :: Int
-    } deriving (Show, Eq)
+  { unGold :: Int
+  }
+  deriving (Show, Eq, Num)
 
 -- | Addition of gold coins.
 instance Semigroup Gold where
-
+  (<>) :: Gold -> Gold -> Gold
+  -- Could just be x + y but writing the full version for the exercise
+  (<>) x y = Gold {unGold = unGold x + unGold y}
 
 instance Monoid Gold where
-
-
-{- | A reward for completing a difficult quest says how much gold
-you'll receive and whether you'll get a special reward.
-
-If you combine multiple rewards, the final reward will contain a
-special prize if at least one of the rewards is special.
--}
+  mempty :: Gold
+  mempty = Gold {unGold = 0}
+
+-- | A reward for completing a difficult quest says how much gold
+-- you'll receive and whether you'll get a special reward.
+--
+-- If you combine multiple rewards, the final reward will contain a
+-- special prize if at least one of the rewards is special.
 data Reward = Reward
-    { rewardGold    :: Gold
-    , rewardSpecial :: Bool
-    } deriving (Show, Eq)
+  { rewardGold :: Gold,
+    rewardSpecial :: Bool
+  }
+  deriving (Show, Eq)
 
 instance Semigroup Reward where
-
+  (<>) :: Reward -> Reward -> Reward
+  (<>) r1 r2 =
+    Reward
+      { rewardGold = rewardGold r1 + rewardGold r2,
+        rewardSpecial = rewardSpecial r1 || rewardSpecial r2
+      }
 
 instance Monoid Reward where
-
-
-{- | 'List1' is a list that contains at least one element.
--}
+  mempty :: Reward
+  mempty =
+    Reward
+      { rewardGold = 0,
+        rewardSpecial = False
+      }
+
+-- | 'List1' is a list that contains at least one element.
 data List1 a = List1 a [a]
-    deriving (Show, Eq)
+  deriving (Show, Eq)
 
 -- | This should be list append.
 instance Semigroup (List1 a) where
+  (<>) :: List1 a -> List1 a -> List1 a
+  (<>) (List1 x xs) (List1 y ys) = List1 x (xs <> (y : ys))
 
+-- | Does 'List1' have the 'Monoid' instance? If no then why?
+-- Answer: List1 does not have a Monoid instance.
+-- The empty element for appending 2 lists is the empty list. List1 being the set of non-empty
+-- lists, [] is not a valid List1. Appending 2 List1 always appends at least 1 element.
+-- instance Monoid (List1 a) where..
```

Comment:
That's correct! 👍🏻 
Or, in other words, there's no such thing as "empty non-empty list" 🙂